### ISSUE TITLE
`ossl_cipher_unpadblock` is not constant-time

### ISSUE THREAD

**Issue by ntauth** [openssl/openssl#16230]
Hello, I was auditing the `ossl_cipher_unpadblock` function in `providers/implementations/ciphers/cipher_commonblock.c` (lines 108-135) and I stumbled across a clear constant-time violation in the unpadding logic. More specifically, AES-CBC decryption does not perform unpadding in constant time. I assume this is known and has not been fixed because of valid reasons. Could you please provide some insights into this?

The version of OpenSSL is `3.0.0-beta2` and the code in question is the following:
```c
int ossl_cipher_unpadblock(unsigned char *buf, size_t *buflen, size_t blocksize)
{
    size_t pad, i;
    size_t len = *buflen;

    if(len != blocksize) {
        ERR_raise(ERR_LIB_PROV, ERR_R_INTERNAL_ERROR);
        return 0;
    }

    /*
     * The following assumes that the ciphertext has been authenticated.
     * Otherwise it provides a padding oracle.
     */
    pad = buf[blocksize - 1];
    if (pad == 0 || pad > blocksize) {
        ERR_raise(ERR_LIB_PROV, PROV_R_BAD_DECRYPT);
        return 0;
    }
    for (i = 0; i < pad; i++) {
       /**
         * This is clearly not constant-time. The loop exits early.
         */
        if (buf[--len] != pad) {
            ERR_raise(ERR_LIB_PROV, PROV_R_BAD_DECRYPT);
            return 0;
        }
    }
    *buflen = len;
    return 1;
}
```

**Comment by slontis** [comment]
I dont think this is new.. in 1.1.1. EVP_DecryptFinal_ex() has equivalent code.

**Comment by ntauth** [comment]
Hi, yes this is indeed not new. 
We were just wondering if this constant time violation was important, and if it isn't, the thought process behind deciding that a constant time violation is not important/vulnerable. Just an honest question that came to mind when auditing the code.

**Comment by slontis** [comment]
@mattcaswell care to comment?

**Comment by mattcaswell** [comment]
I think I just reused the code from 1.1.1. A constant time version would be desirable.

**Comment by ntauth** [comment]
@mattcaswell Thanks a lot! We are currently working on a constant time implementation and we are aiming to submit a PR in the next couple days.

**Comment by ntauth** [comment]
A PR has been opened #16323.

**Comment by nhorman** [comment]
PR Submitted but never updated in response to reviews.

Marking as inactive, to be closed at the end of 3.4 dev barring further input

**Comment by t8m** [comment]
We should try to resurrect the PRs https://github.com/openssl/openssl/pull/16323 and #22777

This looks like an interesting thing to do as security hardening.
